bench_test.go

```diff
@@ -0,0 +1,264 @@
+package yarpc_test
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"testing"
+	"time"
+
+	"github.com/yarpc/yarpc-go"
+	"github.com/yarpc/yarpc-go/encoding/raw"
+	"github.com/yarpc/yarpc-go/transport"
+	yhttp "github.com/yarpc/yarpc-go/transport/http"
+	ytchannel "github.com/yarpc/yarpc-go/transport/tchannel"
+
+	"github.com/stretchr/testify/require"
+	"github.com/uber/tchannel-go"
+	traw "github.com/uber/tchannel-go/raw"
+	"golang.org/x/net/context"
+	"golang.org/x/net/context/ctxhttp"
+)
+
+func yarpcEcho(reqMeta yarpc.ReqMeta, body []byte) ([]byte, yarpc.ResMeta, error) {
+	return body, yarpc.NewResMeta(reqMeta.Context()).Headers(reqMeta.Headers()), nil
+}
+
+var httpEcho http.HandlerFunc = func(w http.ResponseWriter, r *http.Request) {
+	defer r.Body.Close()
+	hs := w.Header()
+	for k, vs := range r.Header {
+		for _, v := range vs {
+			hs.Set(k, v)
+		}
+	}
+
+	_, err := io.Copy(w, r.Body)
+	if err != nil {
+		panic(fmt.Sprintf("HTTP handler failed: %v", err))
+	}
+}
+
+type tchannelEcho struct{ b *testing.B }
+
+func (tchannelEcho) Handle(ctx context.Context, args *traw.Args) (*traw.Res, error) {
+	return &traw.Res{Arg2: args.Arg2, Arg3: args.Arg3}, nil
+}
+
+func (t tchannelEcho) OnError(ctx context.Context, err error) {
+	t.b.Fatalf("request failed: %v", err)
+}
+
+func withDispatcher(b *testing.B, cfg yarpc.Config, f func(yarpc.Dispatcher)) {
+	d := yarpc.NewDispatcher(cfg)
+	require.NoError(b, d.Start(), "failed to start server")
+	defer d.Stop()
+
+	f(d)
+}
+
+func runYARPCClient(b *testing.B, c raw.Client) {
+	reqBody := []byte("hello")
+	for i := 0; i < b.N; i++ {
+		ctx, _ := context.WithTimeout(context.Background(), 100*time.Millisecond)
+		if _, _, err := c.Call(yarpc.NewReqMeta(ctx).Procedure("echo"), reqBody); err != nil {
+			b.Fatalf("request %d failed: %v", i+1, err)
+		}
+	}
+}
+
+func runHTTPClient(b *testing.B, c *http.Client, url string) {
+	reqBody := []byte("hello")
+	for i := 0; i < b.N; i++ {
+		ctx, _ := context.WithTimeout(context.Background(), 100*time.Millisecond)
+		req, err := http.NewRequest("POST", url, bytes.NewReader(reqBody))
+		if err != nil {
+			b.Fatalf("failed to build request %d: %v", i+1, err)
+		}
+
+		req.Header.Add("Rpc-Caller", "http-client")
+		req.Header.Add("Rpc-Encoding", "raw")
+		req.Header.Add("Context-TTL-MS", "100")
+		req.Header.Add("Rpc-Procedure", "echo")
+		req.Header.Add("Rpc-Service", "server")
+
+		res, err := ctxhttp.Do(ctx, c, req)
+		if err != nil {
```

Comment:
I see a mix of `require.NoError` and `if err != nil { b.Fatalf (...) }`. Any reason for using both?